File: backend/src/main/java/observer/quantum/worm/project/ProjectDto.java
package observer.quantum.worm.project;

import io.swagger.v3.oas.annotations.media.Schema;
import java.time.OffsetDateTime;
import lombok.*;
import observer.quantum.worm.user.UserDto;

@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@Schema(name = "Project", description = "Project Data")
public class ProjectDto {
  @Schema(description = "Project ID", example = "1")
  private Long id;

  @Schema(description = "Project slug", example = "project-slug")
  private String slug;

  @Schema(description = "Project title", example = "My Awesome Novel")
  private String title;

  @Schema(description = "Project description", example = "A thrilling adventure story set in space")
  private String description;

  @Schema(description = "User who owns the project")
  private UserDto user;

  @Schema(description = "Date when the project was created", example = "2023-04-15T10:30:00Z")
  private OffsetDateTime createdDate;

  @Schema(description = "Date when the project was last updated", example = "2023-04-16T14:45:00Z")
  private OffsetDateTime updatedDate;

  public ProjectDto(Project project) {
    this.id = project.getId();
    this.title = project.getTitle();
    this.description = project.getDescription();
    this.slug = project.getSlug();
    this.user = new UserDto(project.getUser());
    this.createdDate = project.getCreatedDate();
    this.updatedDate = project.getUpdatedDate();
  }

  public Project toProject() {
    Project project = new Project();
    project.setId(this.id);
    project.setSlug(this.slug);
    project.setTitle(this.title);
    project.setDescription(this.description);
    return project;
  }
}

File: server/src/project/project.dto.ts
import { ApiProperty } from '@nestjs/swagger';
import { ProjectEntity } from './project.entity.js';

export class ProjectDto {
  @ApiProperty({ example: 'my-project-id', required: false })
  id: string;

  @ApiProperty({ example: 'my-project-slug' })
  slug: string;

  @ApiProperty({ example: 'My Awesome Project' })
  title: string;

  @ApiProperty({ example: 'This is a cool project.', required: false })
  description?: string;

  @ApiProperty({example: 'user-1', required: true })
  username: string;

  @ApiProperty({ type: Date, example: '2023-01-01T00:00:00.000Z' })
  createdDate: Date;

  @ApiProperty({ type: Date, example: '2023-01-01T00:00:00.000Z' })
  updatedDate: Date;

  constructor(entity?: ProjectEntity) {
    if (entity) {
      this.slug = entity.slug;
      this.title = entity.title;
      this.description = entity.description;
      this.createdDate = entity.createdDate;
      this.updatedDate = entity.updatedDate;
      this.username = entity.user.username;
    }
  }

  toEntity(): ProjectEntity {
    const project = new ProjectEntity();
    project.slug = this.slug;
    project.title = this.title;
    project.description = this.description;
    project.createdDate = this.createdDate;
    project.updatedDate = this.updatedDate;
    return project;
  }
}
